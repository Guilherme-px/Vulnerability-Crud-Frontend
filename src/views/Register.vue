<template>
    <div class="form">
        <v-flex xs10 sm8 md6 lg5>
            <v-card dark>
                <FormTitle title="Cadastro" />
                <ModalFormMessage
                    :snackbar.sync="snackbar"
                    :message="message"
                    :color="color"
                />
                <v-form ref="form" class="d-flex flex-column">
                    <FormInputs
                        v-model="vulnerabilities.title"
                        label="Título"
                    />
                    <FormInputs
                        v-model="vulnerabilities.description"
                        label="Descrição"
                    />
                    <v-select
                        label="Grau de criticidade"
                        v-model="vulnerabilities.criticality"
                        :items="items.criticality"
                    ></v-select>
                    <v-select
                        label="Tipo de vulnerabilidade"
                        v-model="vulnerabilities.vulnerability"
                        :items="items.vulnerability"
                    ></v-select>
                    <v-file-input @change="onselect"></v-file-input>
                    <FormInputs
                        v-model="vulnerabilities.solution"
                        label="Solução proposta"
                    />
                    <FormBtns :handleSubmit="register" :goBack="goList" />
                </v-form>
            </v-card>
        </v-flex>
    </div>
</template>

<script>
    import FormBtns from '../components/forms/FormBtns.vue';
    import FormInputs from '../components/forms/FormInputs.vue';
    import FormTitle from '../components/forms/FormTitle.vue';
    import ModalFormMessage from '../components/forms/ModalFormMessage.vue';

    import axios from '../utils/axios';

    export default {
        components: {
            FormTitle,
            ModalFormMessage,
            FormInputs,
            FormBtns,
        },
        name: 'Register',
        data: () => ({
            snackbar: false,
            vulnerabilities: {},
            message: '',
            color: '',
            items: {
                criticality: ['low', 'min', 'high'],
                vulnerability: ['DAST', 'SAST', 'NETWORK'],
            },
            imageFile: null,
        }),
        methods: {
            // detects event in the file field and assigns the event result to the evidence field
            onselect(e) {
                this.vulnerabilities.evidence = e;
            },
            // method to call vulnerability registration post route
            async register() {
                // assign formdata to fd
                const fd = new FormData();

                // loop through vulnerabilities keys
                await Object.keys(this.vulnerabilities).forEach((key) => {
                    if (key === 'evidence') {
                        fd.append('evidence', this.vulnerabilities[key]);
                    } else {
                        fd.append(key, this.vulnerabilities[key]);
                    }
                });

                await axios
                    .post('/vulnerability/register', fd, {
                        headers: {
                            'Content-Type': 'multipart/form-data',
                        },
                    })
                    .then(() => {
                        // if everything goes right vulnerabilities are reset
                        this.vulnerabilities = {};
                    })
                    .catch((e) => {
                        // if an error occurs, the error alert is displayed with the error message coming from the api
                        this.message = e.response.data;
                        this.color = 'error';
                        this.snackbar = true;
                    });
            },
            // go to list screen
            goList() {
                this.$router.push({
                    name: 'List',
                });
            },
        },
    };
</script>

<style lang="sass">
    .form
        display: flex
        justify-content: center
        align-items: center
        margin: 50px 0
</style>
