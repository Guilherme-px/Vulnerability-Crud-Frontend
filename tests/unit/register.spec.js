import Vue from 'vue';
import Vuetify from 'vuetify';
Vue.use(Vuetify);

// Utilities
import { createLocalVue, mount } from '@vue/test-utils';

// Components
import Register from '@/views/Register.vue';

describe('Register.vue', () => {
    const localVue = createLocalVue();
    let vuetify;

    beforeEach(() => {
        vuetify = new Vuetify();
    });

    const mountFunction = (options) => {
        return mount(Register, {
            localVue,
            vuetify,
            ...options,
        });
    };

    it('should check if the component exists', () => {
        const wrapper = mountFunction();

        expect(wrapper.exists()).toBe(true);
    });

    it('should check if the component is visible', () => {
        const wrapper = mountFunction();

        expect(wrapper.isVisible()).toBe(true);
    });

    it('should display the registration text in an h2 tag', () => {
        const wrapper = mountFunction({
            propsData: { title: 'Cadastro' },
        });

        const title = wrapper.find('.login__title > h2');

        expect(title.text()).toBe('Cadastro');
    });

    it('should contain 3 text input', () => {
        const wrapper = mountFunction({});

        const textInput = wrapper.findAllComponents({ name: 'FormInput' });
        expect(textInput.length).toEqual(3);
    });

    it('should contain 2 select fields', () => {
        const wrapper = mountFunction({});

        const selectInput = wrapper.findAll('.v-select');
        expect(selectInput.length).toEqual(2);
    });

    it('should contain 1 file fields', () => {
        const wrapper = mountFunction({});

        const fileInput = wrapper.findAll('.v-file-input__text');
        expect(fileInput.length).toEqual(1);
    });

    it('should contain an input with the title label', () => {
        const wrapper = mountFunction({
            propsData: { label: 'Título' },
        });

        const label = wrapper.findAll('.v-label');
        expect(label.at(0).text()).toBe('Título');
    });

    it('should contain an input with the description label', () => {
        const wrapper = mountFunction({
            propsData: { label: 'Descrição' },
        });

        const label = wrapper.findAll('.v-label');
        expect(label.at(1).text()).toBe('Descrição');
    });

    it('should contain an input with the label Criticality degree', () => {
        const wrapper = mountFunction({
            propsData: { label: 'Grau de criticidade' },
        });

        const label = wrapper.findAll('.v-label');
        expect(label.at(2).text()).toBe('Grau de criticidade');
    });

    it('should contain an entry labeled Vulnerability Type', () => {
        const wrapper = mountFunction({
            propsData: { label: 'Tipo de vulnerabilidade' },
        });

        const label = wrapper.findAll('.v-label');
        expect(label.at(3).text()).toBe('Tipo de vulnerabilidade');
    });

    it('should contain an input with the label Proposed solution', () => {
        const wrapper = mountFunction({
            propsData: { label: 'Solução proposta' },
        });

        const label = wrapper.findAll('.v-label');
        expect(label.at(4).text()).toBe('Solução proposta');
    });

    it('should a button component', () => {
        const wrapper = mountFunction();

        const button = wrapper.findComponent({ name: 'FormBtns' });

        expect(button.exists()).toBe(true);
    });

    it('should have a visible button component', () => {
        const wrapper = mountFunction();

        const button = wrapper.findComponent({ name: 'FormBtns' });

        expect(button.isVisible()).toBe(true);
    });

    it('should contain a button named save and one named back', () => {
        const wrapper = mountFunction();

        const button = wrapper.findAll('.v-btn');

        expect(button.at(0).text()).toBe('Salvar');
        expect(button.at(1).text()).toBe('Voltar');
    });
});
