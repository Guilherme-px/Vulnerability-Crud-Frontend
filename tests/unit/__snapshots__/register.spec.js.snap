// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelloWorld.vue should have a custom title and match snapshot 1`] = `
<div class="form" title="Foobar">
  <v-flex xs10="" sm8="" md6="" lg5="">
    <v-card dark="">
      <div class="login__title">
        <h2>Cadastro</h2>
      </div>
      <v-snackbar timeout="4000" top="" right="" color="" class="my-15">
        <v-icon dark="" large="" class="pr-3">mdi-alert-circle-outline</v-icon>
        Atenção, os seguintes campos são obrigatórios:
        <div>

        </div>
      </v-snackbar>
      <v-form class="d-flex flex-column">
        <div>
          <v-text-field label="Título" value=""></v-text-field>
        </div>
        <div>
          <v-text-field label="Descrição" value=""></v-text-field>
        </div>
        <div>
          <v-select label="Grau de criticidade" items="low,min,high"></v-select>
        </div>
        <div>
          <v-select label="Tipo de vulnerabilidade" items="DAST,SAST,NETWORK"></v-select>
        </div>
        <v-file-input></v-file-input>
        <div>
          <v-text-field label="Solução proposta" value=""></v-text-field>
        </div>
        <div class="form__btn--group">
          <v-btn color="primary mt-10 mb-10" x-large="" class="form__btn">Salvar</v-btn>
          <v-btn color="primary mt-10 mb-10" x-large="" class="form__btn">Voltar</v-btn>
        </div>
      </v-form>
    </v-card>
  </v-flex>
</div>
`;

exports[`Register.vue should  1`] = `
<div class="form" label="Titulo">
  <div class="flex xs10 sm8 md6 lg5">
    <div class="v-card v-sheet theme--dark">
      <div class="login__title">
        <h2>Cadastro</h2>
      </div>
      <div class="v-snack my-15 v-snack--has-background v-snack--right v-snack--top">
        <transition-stub name="v-snack-transition">
          <div class="v-snack__wrapper v-sheet theme--dark" style="display: none;">
            <div role="status" aria-live="polite" class="v-snack__content"><i aria-hidden="true" class="v-icon notranslate pr-3 mdi mdi-alert-circle-outline theme--dark" style="font-size: 36px;"></i>
              Atenção, os seguintes campos são obrigatórios:
              <div>

              </div>
            </div>
            <div class="v-snack__action "></div>
          </div>
        </transition-stub>
      </div>
      <form novalidate="novalidate" class="v-form d-flex flex-column">
        <div>
          <div class="v-input theme--dark v-text-field">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-50" class="v-label theme--dark" style="left: 0px; position: absolute;">Título</label><input id="input-50" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--dark">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div class="v-input theme--dark v-text-field">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-54" class="v-label theme--dark" style="left: 0px; position: absolute;">Descrição</label><input id="input-54" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--dark">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div class="v-input theme--dark v-text-field v-select">
            <div class="v-input__control">
              <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-58" class="v-input__slot">
                <div class="v-select__slot"><label for="input-58" class="v-label theme--dark" style="left: 0px; position: absolute;">Grau de criticidade</label>
                  <div class="v-select__selections"><input id="input-58" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                  </div><input type="hidden">
                </div>
                <div class="v-menu">
                  <!---->
                </div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--dark">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div class="v-input theme--dark v-text-field v-select">
            <div class="v-input__control">
              <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-64" class="v-input__slot">
                <div class="v-select__slot"><label for="input-64" class="v-label theme--dark" style="left: 0px; position: absolute;">Tipo de vulnerabilidade</label>
                  <div class="v-select__selections"><input id="input-64" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                  </div><input type="hidden">
                </div>
                <div class="v-menu">
                  <!---->
                </div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--dark">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="v-input theme--dark v-text-field v-file-input">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend"><button type="button" aria-label="prepend icon" class="v-icon notranslate v-icon--link mdi mdi-paperclip theme--dark"></button></div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot">
                <div class="v-file-input__text"></div><input id="input-69" type="file">
              </div>
              <div class="v-input__append-inner">
                <div></div>
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--dark">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div class="v-input theme--dark v-text-field">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-74" class="v-label theme--dark" style="left: 0px; position: absolute;">Solução proposta</label><input id="input-74" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--dark">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="form__btn--group"><button type="button" class="form__btn v-btn v-btn--is-elevated v-btn--has-bg theme--dark v-size--x-large primary mt-10 mb-10"><span class="v-btn__content">Salvar</span></button> <button type="button" class="form__btn v-btn v-btn--is-elevated v-btn--has-bg theme--dark v-size--x-large primary mt-10 mb-10"><span class="v-btn__content">Voltar</span></button></div>
      </form>
    </div>
  </div>
</div>
`;
